applications/FemToDemApplication/python_scripts/CouplingFemDem3D.py

```diff
@@ -0,0 +1,1035 @@
+
+			if NumberOfActiveElements == 0 and node.GetValue(KratosFemDem.INACTIVE_NODE) == False:
+
+				Id = node.Id
+
+				DEMnode = self.SpheresModelPart.GetNode(Id)
+				node.SetValue(KratosFemDem.INACTIVE_NODE, True)
+				node.Set(KratosMultiphysics.TO_ERASE, True) # added
+				DEMnode.SetValue(KratosFemDem.INACTIVE_NODE, True)
+				DEMnode.Set(KratosMultiphysics.TO_ERASE, True)
+
+			# Reset the value to the next step
+			node.SetValue(KratosFemDem.NUMBER_OF_ACTIVE_ELEMENTS, 0)
+
+		# Remove inactive nodes
+		self.SpheresModelPart.RemoveElementsFromAllLevels(KratosMultiphysics.TO_ERASE)
+		self.FEM_Solution.main_model_part.GetRootModelPart().RemoveNodesFromAllLevels(KratosMultiphysics.TO_ERASE) # added
+
+#============================================================================================================================
+	def UpdateDEMVariables(self):
+
+		FEM_Nodes = self.FEM_Solution.main_model_part.Nodes
+		DEM_Nodes = self.SpheresModelPart.Nodes
+
+		for DEM_Node in DEM_Nodes:  # Loop over DEM nodes
+
+			if (DEM_Node.GetValue(KratosFemDem.INACTIVE_NODE) == False):
+
+				Id = DEM_Node.Id
+				Corresponding_FEM_Node = self.FEM_Solution.main_model_part.GetNode(Id)
+
+				Coordinates    = self.GetNodeCoordinates(Corresponding_FEM_Node)
+				Velocity_x     = Corresponding_FEM_Node.GetSolutionStepValue(KratosMultiphysics.VELOCITY_X)
+				Velocity_y     = Corresponding_FEM_Node.GetSolutionStepValue(KratosMultiphysics.VELOCITY_Y)
+				Velocity_z     = Corresponding_FEM_Node.GetSolutionStepValue(KratosMultiphysics.VELOCITY_Z)
+				Displacement_x = Corresponding_FEM_Node.GetSolutionStepValue(KratosMultiphysics.DISPLACEMENT_X)
+				Displacement_y = Corresponding_FEM_Node.GetSolutionStepValue(KratosMultiphysics.DISPLACEMENT_Y)
+				Displacement_z = Corresponding_FEM_Node.GetSolutionStepValue(KratosMultiphysics.DISPLACEMENT_Z)
+
+				# Update Coordinates
+				DEM_Node.X = Coordinates[0]
+				DEM_Node.Y = Coordinates[1]
+				DEM_Node.Z = Coordinates[2]
+
+				# Update Displacements
+				DEM_Node.SetSolutionStepValue(KratosMultiphysics.DISPLACEMENT_X, Displacement_x)
+				DEM_Node.SetSolutionStepValue(KratosMultiphysics.DISPLACEMENT_Y, Displacement_y)
+				DEM_Node.SetSolutionStepValue(KratosMultiphysics.DISPLACEMENT_Z, Displacement_z)
+
+				# Update Velocities
+				DEM_Node.SetSolutionStepValue(KratosMultiphysics.VELOCITY_X, Velocity_x)
+				DEM_Node.SetSolutionStepValue(KratosMultiphysics.VELOCITY_Y, Velocity_y)
+				DEM_Node.SetSolutionStepValue(KratosMultiphysics.VELOCITY_Z, Velocity_z)
+
+#============================================================================================================================
+	def TransferNodalForcesToFEM(self):
+
+		DEM_Nodes = self.SpheresModelPart.Nodes
+
+		for DEM_node in DEM_Nodes:
+
+			# Get the contact forces from the DEM
+			Id = DEM_node.Id
+			Force_X = DEM_node.GetSolutionStepValue(KratosMultiphysics.TOTAL_FORCES_X)

```

Comment:
Use the vector, you are accessing 3 times to the database, when you can do it once